src/libstd/fs.rs

```diff
@@ -246,8 +246,12 @@ pub struct DirBuilder {
 /// ```
 #[unstable(feature = "fs_read_write", issue = "46588")]
 pub fn read<P: AsRef<Path>>(path: P) -> io::Result<Vec<u8>> {
-    let mut bytes = Vec::new();
-    File::open(path)?.read_to_end(&mut bytes)?;
+    // Allocate one extra byte so the buffer doesn't need to grow before the final `read` call at
+    // the end of the file.  Don't worry about `usize` overflow because this will fail anyway if
+    // the file doesn't fit into memory.
+    let mut bytes = Vec::with_capacity(metadata(&path)?.len() as usize + 1);
```

Comment:
Are there contexts/weird files where the metadata lookup won't work but the file read will?

Reply:
I'm not totally certain.  I based this on https://github.com/rust-lang/rust/pull/45928#issuecomment-347638206:

> You can always return 0 if something went wrong, but it seems like it's nice to not throw away those errors. I kind of doubt that reads would succeed if you can't stat the file anymore.

but I do think it would be reasonable to just start with an empty buffer if the stat call fails.

Reply:
I agree that starting with an empty buffer if the stat call fails is probably the right thing to do. It's potentially worth pointing out that it seems like this may not be faster (https://github.com/rust-lang/rust/pull/35844#issuecomment-241271162) based on past benchmarks so maybe we shouldn't do this.

Reply:
I think those numbers may no longer be accurate - we no longer zero the buffer before reading into it for Files, and read_to_end has been tweaked a bit to behave differently. #45928 has some benchmarks.